### PR TITLE
Extend the PublicKeyInfo proto definition with additional evidence info

### LINKED ISSUE #3656: Update Oak proto to include a full enclave evidence
Our current Oak interface between the Client and the Launcher doesn't include all of the things necessary for performing a Client-Side Attestation Verification (https://github.com/project-oak/oak/issues/3641).

The structure (that should be probably renamed to `Evidence`) looks like this:
https://github.com/project-oak/oak/blob/d8c147476db02e4360ae06ba19c1363c5260fe13/oak_remote_attestation_noninteractive/proto/v1/messages.proto#L24-L31

And in order to do the attestation verification we need the following things to be sent to the client:
- Enclave public key
- Enclave config
- Attestation report
- AMD certificates
- Endorcement
- Rekor inclusion proof

Ref https://github.com/project-oak/oak/issues/3442

### PR DESCRIPTION
**Pull request by rbehjati**
Fixes #3656

@project-oak/core Putting this out early for discussion. 

1. What do you think about the naming?
2. Any important points that could be included in the documentation at this stage?
3. Is there any other piece of information that should be included in `PublicKeyInfo`?

### CODE REVIEW

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto [hunk 1]
```diff
@@ -24,10 +24,23 @@ option java_package = "oak.session.noninteractive.v1";
...
+  // The serialized AMD-SEV-SNP certificate.
+  bytes tee_certificate = 3;
+  // The serialized binary attestation evidence.
+  bytes binary_attestation = 4;
+  // The serialized server configuration. The details of the configuration is application-specific.
+  // This field is optional and can be left empty.
+  bytes server_config = 5;
+}
+
+message BinaryAttestation {
```
Comment (conradgrobler) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:39]: How is this related to `binary_attestation` above?

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto [hunk 2]
```diff
@@ -24,10 +24,23 @@ option java_package = "oak.session.noninteractive.v1";
 message PublicKeyInfo {
```
Comment (conradgrobler) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:24]: I assume this is only the message between the untrusted launcher and the guest, is that correct? The trusted part will not have most of these fields.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto [hunk 3]
```diff
@@ -24,10 +24,23 @@ option java_package = "oak.session.noninteractive.v1";
...
   // The serialized enclave public key.
   // TODO(#3442): Specify key format.
-  bytes public_key = 1;
+  bytes enclave_public_key = 1;
```
Comment (conradgrobler) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:27]: I think we need an additional field to specify the intended use of the public key. It is usually important to know whether a key is used for signing or encryption. Perhaps we should always create two keys, one signing key and one encryption key.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto [hunk 4]
```diff
@@ -24,10 +24,23 @@ option java_package = "oak.session.noninteractive.v1";
...
   // The serialized attestation report binding the public key.
   bytes attestation = 2;
-  // TODO(#3442): Add evidence (endorsements and inclusion proofs) for the attested binary.
+  // The serialized AMD-SEV-SNP certificate.
+  bytes tee_certificate = 3;
```
Comment (conradgrobler) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:31]: This is actually three different certificates packaged in two different files for AMD SEV-SNP.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto [hunk 5]
```diff
@@ -24,10 +24,23 @@ option java_package = "oak.session.noninteractive.v1";
...
+  // The serialized binary attestation evidence.
+  bytes binary_attestation = 4;
+  // The serialized server configuration. The details of the configuration is application-specific.
+  // This field is optional and can be left empty.
+  bytes server_config = 5;
```
Comment (conradgrobler) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:36]: I assume this config must be bound to the attestation report in some way, so either this must be signed using the enclave_public_key (if it is a signing key), or its hash should apear in the attestation report's additional data. We should decide which of these approaches to follw and be clearer in the comment on how we prove the integrity of this field.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 2 above)
Comment (ipetr0v) [inline oak_remote_attestation_noninteractive/proto/v1/messages.proto:24]: Do you think that it could be called `Evidence` or `AttestationEvidence` ?

Comment (rbehjati) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:24]: Are you suggesting this to match the RATS terminology? 

I'd be happy to rename it to `AttestationEvidence`, if the `AttestationEvidence` in the RATS terminology is actually meant to contain this information and be exchanged between an untrusted component and the client.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 3 above)
Comment (rbehjati) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:27]: My understanding from the description in https://github.com/project-oak/oak/blob/main/docs/remote-attestation.md is that this key is for encryption. Are there any scenarios where the key would be used for signing?

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 4 above)
Comment (rbehjati) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:31]: I'll add this to the comment. Do you have a reference I could include in the comment?

Perhaps I could define a separate struct for it here. WDYT?

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 5 above)
Comment (rbehjati) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:36]: Good point. The last thing I remember is that the config hash should be included in the attestation report's additional data. I'll update the comment accordingly.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 1 above)
Comment (rbehjati) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:39]: Good catch. This is the correct type of `binary_attestation` above. I'll update.

**Review by rbehjati — COMMENTED** [review]
Thanks for the comments :)

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 4 above)
Comment (conradgrobler) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:31]: The information specific to AMD SEV-SNP is in https://www.amd.com/system/files/TechDocs/57230.pdf.

I don't think this message will only be used form AMD SEV-SNP, so perhaps the field can be renamed to indicate that it contains multiple certificates and add a comment that the details of how the certificates are encoded in this byte array are implementation-specific.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 3 above)
Comment (conradgrobler) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:27]: If we want to sign some future content to prove that it came from the TEE we need a signing key. An example would be the future plans to implement an interactive handshake with an attestation verification service that will allow the TEE to prove its identity. The TEE would generate an ephemeral key pair as part of the handshake and then must be able to prove to the verification service that the ephemeral key pair was actually generated in the TEE.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 5 above)
Comment (conradgrobler) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:36]: For now I think that is fine. If we add support in future for a signing key we can perhaps use that to sign the server config and any additional data from the TEE.

File: oak_remote_attestation_noninteractive/proto/v1/messages.proto (hunk 2 above)
Comment (ipetr0v) [reply oak_remote_attestation_noninteractive/proto/v1/messages.proto:24]: yeah
It's not only the Evidence
In RATS it will be a combination of `Evidence` and `Endorcements`
I think we can rename it into `AttestationEvidence` for now, until we decide whether we want to use the term Endorcements from RATS or Transparent Release

**Review by rbehjati — COMMENTED** [review]
Thanks for the comments. I've updated the proto accordingly, and have added a simple integration test.

**Review by conradgrobler — APPROVED** [review]
